dev/tools/localization/gen_l10n.dart

```diff
@@ -0,0 +1,412 @@
+  String message = bundle[key];
+  for (String placeholder in placeholders)
+    message = message.replaceAll('{$placeholder}', '#$placeholder#');
+
+  final Map<String, String> pluralIds = <String, String>{
+    '=0': 'zero',
+    '=1': 'one',
+    '=2': 'two',
+    'few': 'few',
+    'many': 'many',
+    'other': 'other'
+  };
+
+  final List<String> methodArgs = <String>[
+    ...placeholders,
+    'locale: _localeName',
+    ...genIntlMethodArgs(bundle, key),
+  ];
+
+  for(String pluralKey in pluralIds.keys) {
+    final RegExp expRE = RegExp('($pluralKey){([^}]+)}');
+    final RegExpMatch match = expRE.firstMatch(message);
+    if (match.groupCount == 2) {
+      String argValue = match.group(2);
+      for (String placeholder in placeholders)
+        argValue = argValue.replaceAll('#$placeholder#', '\$$placeholder');
+
+      methodArgs.add("${pluralIds[pluralKey]}: '$argValue'");
+    }
+  }
+
+  return pluralMethodTemplate
+    .replaceAll('@methodName', key)
+    .replaceAll('@methodParameters', genMethodParameters(bundle, key, 'int').join(', '))
+    .replaceAll('@intlMethodArgs', methodArgs.join(',\n      '));
+}
+
+String genSupportedLocaleProperty(Set<LocaleInfo> supportedLocales) {
+  const String prefix = 'static const List<Locale> supportedLocales = <Locale>[ \n    Locale(''';
+  const String suffix = '),\n  ];';
+
+  String resultingProperty = prefix;
+  for (LocaleInfo locale in supportedLocales) {
+    final String languageCode = locale.languageCode;
+    final String countryCode = locale.countryCode;
+
+    resultingProperty += '\'$languageCode\'';
+    if (countryCode != null)
+      resultingProperty += ', \'$countryCode\'';
+    resultingProperty += '),\n    Locale(';
+  }
+  resultingProperty = resultingProperty.substring(0, resultingProperty.length - '),\n    Locale('.length);
+  resultingProperty += suffix;
+
+  return resultingProperty;
+}
+
+bool _isValidClassName(String className) {
+  // Dart class name cannot contain non-alphanumeric symbols
+  if (className.contains(RegExp(r'[^a-zA-Z\d]')))
+    return false;
+  // Dart class name must start with upper case character
+  if (className[0].contains(RegExp(r'[a-z]')))
+    return false;
+  // Dart class name cannot start with a number
+  if (className[0].contains(RegExp(r'\d')))
+    return false;
+  return true;
+}
+
+bool _isDirectoryReadableAndWritable(String statString) {
+  if (statString[0] == '-' || statString[1] == '-')
+    return false;
+  return true;
+}
+
+String _importFilePath(String path, String fileName) {
+  final String replaceLib = path.replaceAll('lib/', '');
+  return '$replaceLib/$fileName.dart';
+}
+
+Future<void> main(List<String> args) async {
+  final argslib.ArgParser parser = argslib.ArgParser();
+  parser.addOption('arb-dir', defaultsTo: path.join('lib', 'l10n'));
+  parser.addOption('template-arb-file', defaultsTo: 'app_en.arb');
+  parser.addOption('output-localization-file', defaultsTo: 'app_localizations');
```

Comment:
It might be less confusing if this defaulted to `app_localizations.dart`; if we didn't add the .dart suffix to the output file parameter

Reply:
Done